feat(solana): add Solana support

gabrielKerekes opened the pull request:
We had multiple calls on this topic so I will not write a long description. If there are questions, please just ask them during the review. We can also schedule another call. 

The last commit (https://github.com/trezor/trezor-firmware/commit/c76de12d76f328047ce146a60757473e54a9c9b2) adds simple tests for all the instructions just so we can be sure that tx signing still works. We will add tests properly once the discussion around their implementation is finalised - see [open comment here](https://github.com/trezor/trezor-firmware/pull/3318#issuecomment-1759185698).

We'll generate UI test fixtures after the review process is finalised.

Connect implementation can be found here: https://github.com/trezor/trezor-suite/pull/9655

Useful resources for review:
- [here's a repo](https://github.com/gabrielKerekes/solana-transactions) you can clone to build various Solana transactions for testing (the same transactions can be found in `sign_tx.tmp.json`)
- [Solana docs](https://docs.solana.com/)
- Programs docs:
  - [System program](https://docs.rs/solana-program/latest/solana_program/system_instruction/enum.SystemInstruction.html)
  - [Stake program](https://docs.rs/solana-sdk/latest/solana_sdk/stake/instruction/enum.StakeInstruction.html)
  - [Token program](https://docs.rs/spl-token/latest/spl_token/instruction/enum.TokenInstruction.html)
  - [Associated Token Account Program](https://docs.rs/spl-associated-token-account/latest/spl_associated_token_account/instruction/enum.AssociatedTokenAccountInstruction.html)
  - [Compute budget program](https://github.com/solana-labs/solana/blob/master/sdk/src/compute_budget.rs#L25) + [general docs](https://docs.solana.com/transaction_fees#prioritization-fee)
  - [Memo program](https://spl.solana.com/memo)
- [UI proposal doc](https://docs.google.com/spreadsheets/d/1gYK48kZWtMkHEaopQho45B0kJE5eNlUa19Lh0aX8nKM/edit#gid=0)

matejcik commented on python/src/trezorlib/cli/solana.py:
```diff
@@ -0,0 +1,56 @@
+
+@cli.command()
+@click.option("-n", "--address", required=True, help=PATH_HELP)
+@click.option("-t", "--serialized-tx", required=True)
```
change to `@click.argument`

and honestly maybe add a default for `-n` so that `trezorctl solana sign-tx LAKSJFALKSJF"LASJ` just works? same for `get-address`

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/d09243c566a1fde90faeaa22c1a87eca740d43ef

matejcik reviewed (commented):
separate review for the implementation part, a review for tests is upcoming

@grdddj please also take a look at this code, especially the UI part where you might be able to make better suggestions

grdddj commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+from trezor.utils import BufferReader
+from trezor.wire import ProcessError
+
+from .parse import parse_property
```
This import and the two above seems to be used only in one function - then it makes sense to import it locally in that function.

grdddj commented on tests/device_tests/solana/construct/custom_constructs.py:
```diff
@@ -0,0 +1,103 @@
+from base58 import b58decode, b58encode
```
When possible, please also provide type-hints to these helper functions (no need to describe some complex objects, but things like `bytes`, `str` or `int` are quick and easy to add)

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/27c9271891a38edf27374989bd952b5065034b19.

matejcik replied:
on my system `from base58` is not recognized.

however we have base58 encoding in `trezorlib.tools`

gabrielKerekes replied:
I'm not sure exactly what the problem is but decoding `11111111111111111111111111111111` with base58 from trezorlib yields 33 bytes while decoding with the `base58` package yields 32 bytes.

I expect it has to do with [this line](https://github.com/trezor/trezor-firmware/blob/02e6766e2daabb0099c954ecba4a0cd89673aa20/python/src/trezorlib/tools.py#L181) and `nPad` being 1 but I don't know the specifics of base58 so I don't know why it's there or what might be wrong.

Actually with Trezorlib base58 `b58encode(b58decode("11111111111111111111111111111111"))` does not equal `"11111111111111111111111111111111"`. (there's the extra byte)

matejcik replied:
well that's fun! this should help: 0bec4113b051bcab056d1208d1622e5d202a18a9

gabrielKerekes replied:
Thanks! Cherry picked and used in https://github.com/trezor/trezor-firmware/pull/3359/commits/65397e0bc4830070a1b57c06d8e727dc97778d82.

grdddj commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+            _type = property_template.type
+
+            if _type == "authority":
+                if signer_public_key == value:
```
Could be one-line `if X and Y` instead of nested `if`s

albi84 replied:
[Changes](https://github.com/trezor/trezor-firmware/commit/0e1b2f94c0461a1f18f665d6154d683fe97cf0da)

grdddj commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+
+
+def get_address_type(address_type: int) -> str:
+    if ADDRESS_SIG == address_type:
```
It seems strange that the `address_type` is the second one in comparison (I would expect it to be first always) - it took little more time to realize this is actually `switch`

albi84 replied:
[Changes](https://github.com/trezor/trezor-firmware/commit/41116f64b427dd88d38fd53ab9b26357c65334d7)

grdddj commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+                if signer_public_key == account_value[0]:
+                    continue
+
+            account_data = []
```
Could be type-hinted, just to be more safe

albi84 replied:
[Changes](https://github.com/trezor/trezor-firmware/commit/236b05ecea4131347737d54350c5408d7a082b19)

grdddj commented on core/src/apps/solana/transaction/programs.json:
```diff
@@ -0,0 +1,3228 @@
+{
```
We are just currently implementing translations into firmware, and this file has quite a lot of hardcoded English strings (which will be shown to user).

It is an open question of how (and if - @Hannsek) we will support multiple languages for `Solana`

Do you have some ideas?

Hannsek replied:
I believe we should support it the same as other chains. So the less hardcoded English strings the better.

gabrielKerekes replied:
Is there currently any action required on this from our side?

matejcik replied:
we will need to replace the strings here with translation symbols and make them part of the master json.
(nothing to do for this PR though)

grdddj reviewed (commented):
I have a couple of comments, mostly from code and style POV. Most of them are valid for the whole PR, not just the file where I commented on them. 

The majority is type-hinting and import rules, which we use.

I have not studied the logic side of things much - just read the README.

I have pushed the branch into our repo, so we can see `CI` and some test failures (but looks like only not-recorder `fixtures.json`):
https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/1047266679

The master-diff seems to be quite big and unrelated, some `Cardano` screens changed - maybe it just needs rebase to `main` - https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/5360278876/artifacts/test_ui_report/master_diff.html

albi84 commented on core/src/apps/solana/types.py:
```diff
@@ -0,0 +1,62 @@
+class InstructionIdFormat:
+    length: int
+    is_included_if_zero: bool
+
```
[Changes](https://github.com/trezor/trezor-firmware/commit/77b90c389f378cb5786e83fb2791ff21d8cad613)

albi84 commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    pass
+
+
+def _format_property(value: Any, type: str) -> str | bytes | None:
```
[Changes](https://github.com/trezor/trezor-firmware/commit/284cbcb695e5d962ddbf5270675a63e19620c5f2)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+class Instruction:
+    PROGRAM_ID: str
+    INSTRUCTION_ID: int
+
```
[Changes](https://github.com/trezor/trezor-firmware/commit/1081d13000b945e60e2026bc3204ad614b9e8526)

albi84 commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    # only a single import under the `if` statement, the statement would be empty which would result
+    # in an invalid syntax error.
+    pass
+
```
[Changes](https://github.com/trezor/trezor-firmware/commit/ee8019fd2ba7887ab6313c60143758987fe9c8d2)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+    ui_name: str
+
+    ui_parameter_list: list[str] | None = None
+    ui_account_list: list[tuple[str, Account]] | None = None
```
[Changes](https://github.com/trezor/trezor-firmware/commit/9983a7bb5b07c2ffc9a0255c962b58765ce8e809)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+                parse_property(
+                    reader, property_template.type, property_template.optional
+                ),
+            )
```
[Changes part 1](https://github.com/trezor/trezor-firmware/commit/adb538cebf16051b9680bd96a1b80f5a420ca1e8)

albi84 replied:
[Changes part 2](https://github.com/trezor/trezor-firmware/commit/794a455eb56532a6d06779f5b41528bf95546c3f)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+
+    def set_account(self, account: str, value: Account) -> None:
+        assert self.parsed_accounts is not None
+        self.parsed_accounts[account] = value
```
[Changes](https://github.com/trezor/trezor-firmware/pull/3359#discussion_r1368402260)

albi84 commented on core/src/apps/solana/transaction/parse.py:
```diff
@@ -0,0 +1,227 @@
+def parse_header(serialized_tx: BufferReader) -> tuple[bool, int, int, int, int]:
+    assert_cond(serialized_tx.remaining_count() > 3)
+
+    isLegacy: bool = True
```
[Changes](https://github.com/trezor/trezor-firmware/commit/724c66f6ff1beb3f0a32513d79b49d1b6022ad02)

albi84 commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,121 @@
+    required_signers_count = 0
+
+    is_legacy: bool
+    version: int
```
[Changes](https://github.com/trezor/trezor-firmware/commit/3da5bc39d5b2d53b08cc730db64ffc6c1b763532)

albi84 commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    if type in ("pubkey", "authority"):
+        return base58.encode(value)
+    elif type == "lamports":
+        return f"{value / 10 ** 9} SOL"
```
[Changes](https://github.com/trezor/trezor-firmware/commit/40039407cf6d6fa4c4ecfd49c2c8596fa919b8fe)

albi84 commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,121 @@
+    instructions: list[Instruction] = []
+
+    address_lookup_tables_rw_addresses: list[AddressReference] = []
+    address_lookup_tables_ro_addresses: list[AddressReference] = []
```
[Changes](https://github.com/trezor/trezor-firmware/commit/3ae68218a88d9aa0691df623f2a2c098f2a5611b)

albi84 commented on core/src/apps/solana/sign_tx.py:
```diff
@@ -0,0 +1,142 @@
+
+    base_fee = SOLANA_BASE_FEE_LAMPORTS * number_of_signers
+
+    unit_price: int | None = None
```
[Changes](https://github.com/trezor/trezor-firmware/commit/752faefaf98b69058fda1f12b0db9ae4caed72a6)

albi84 commented on core/src/apps/solana/sign_tx.py:
```diff
@@ -0,0 +1,142 @@
+    base_fee = SOLANA_BASE_FEE_LAMPORTS * number_of_signers
+
+    unit_price: int | None = None
+    unit_limit: int = SOLANA_COMPUTE_UNIT_LIMIT
```
[Changes](https://github.com/trezor/trezor-firmware/commit/5dc6476a0a72a4d570b6427f9c828440e313c8ac)

gabrielKerekes commented on common/protob/messages-solana.proto:
```diff
@@ -0,0 +1,61 @@
+ */
+message SolanaGetAddress {
+    repeated uint32 address_n = 1;  // BIP-32 path to derive the key from master node
+    optional bool show_display = 2; // optionally show on display before sending the result
```
Added to `GetAddress` here -> https://github.com/trezor/trezor-firmware/pull/3359/commits/dd6058f533ad1ac07300169a3f470be6973b125a I'd rather not add it to sign tx since the UI is rather generic - but if you insist, we could probably handle it for all addresses.

gabrielKerekes commented:
https://github.com/trezor/trezor-firmware/pull/3359/commits/0db7dc4d976303aa32e359e826d7a7460f5b3503 and https://github.com/trezor/trezor-firmware/pull/3359/commits/1f0b74bee2d04a4e7eb30bf971f7259dc9ebd38e added based on feedback from QA.

albi84 commented on core/src/apps/solana/transaction/instructions.py:
```diff
@@ -0,0 +1,5878 @@
+            "11111111111111111111111111111111",
+            5,
+        ),
+        "SystemProgramInitializenonceaccountInstruction": (
```
[Changes](https://github.com/trezor/trezor-firmware/commit/7fc3d898cdb7acc630ebe0df8bbb37bc6effe674)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+
+        if not is_instruction_supported:
+            self.instruction_data = reader.read()
+            self.accounts = accounts
```
[Changes](https://github.com/trezor/trezor-firmware/commit/43609f5661758aee0cd6815a96112214138ce6ff)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+
+        if supports_multisig and len(accounts) > len(accounts_template):
+            self.is_multisig = True
+            self.multisig_signers = accounts[len(accounts_template) :]
```
[Changes](https://github.com/trezor/trezor-firmware/commit/816defa398c36fb273a2e9d6f280d023318b0ad0)

gabrielKerekes commented on python/src/trezorlib/cli/solana.py:
```diff
@@ -0,0 +1,56 @@
+if TYPE_CHECKING:
+    from ..client import TrezorClient
+
+PATH_HELP = "BIP-32 path to key, e.g. m/44'/501'/0'"
```
Actually `'` is used also in ethereum, cardano, monero, ripple and others. So I'd suggest keeping it as is and if it should be updated then the other blockchains should be updated as well.

matejcik replied:
[way ahead of you](https://github.com/trezor/trezor-firmware/issues/3037)

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/d09243c566a1fde90faeaa22c1a87eca740d43ef

gabrielKerekes commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+        raise ValueError(f"Account {account_name} not found")
+
+    @classmethod
+    def is_type_of(cls, ins: Any) -> TypeGuard["Instruction"]:
```
https://github.com/trezor/trezor-firmware/pull/3359/commits/89ca10a6619a3bbc089e27591112179e75f2333c

gabrielKerekes commented on core/src/apps/solana/transaction/instructions.py.mako:
```diff
@@ -0,0 +1,232 @@
+                ins.program_id == cls.PROGRAM_ID
+                and ins.instruction_id == cls.INSTRUCTION_ID
+            )
+
```
https://github.com/trezor/trezor-firmware/pull/3359/commits/89ca10a6619a3bbc089e27591112179e75f2333c

gabrielKerekes commented on core/src/apps/solana/transaction/instructions.py.mako:
```diff
@@ -0,0 +1,232 @@
+    class ${getClassName(program, instruction)}(Instruction):
+        PROGRAM_ID = ${getProgramId(program)}
+        INSTRUCTION_ID = ${getInstructionIdText(program, instruction)}
+
```
https://github.com/trezor/trezor-firmware/pull/3359/commits/89ca10a6619a3bbc089e27591112179e75f2333c

gabrielKerekes commented on core/tools/build_solana_templates:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
I did that in https://github.com/trezor/trezor-firmware/pull/3359/commits/59acc63af220d31b0af38824f954f75709e54561. But shouldn't the timestamp be set to either `programs.json` or `instructions.mako.py` depending on which is newer? (I'm not sure why this is needed so I can't tell)

matejcik replied:
this is so that the build process knows that the generated instructions.py haven't changed since last build.
and yes, it would be a bit better to do best of `programs.json` and the `.mako` file, so that we can apply the same optimization to running `build_solana_templates` itself.

gabrielKerekes replied:
Updated to the most recent timestamp in https://github.com/trezor/trezor-firmware/pull/3359/commits/2d77f04cf0bb67d41fc927d54b2183d18e641fb0.

gabrielKerekes commented on core/SConscript.firmware:
```diff
@@ -72,6 +72,8 @@ CPPDEFINES_MOD += [
     ('USE_CARDANO', '1' if EVERYTHING else '0'),
     ('USE_NEM', '1' if (EVERYTHING and TREZOR_MODEL != "R") else '0'),
     ('USE_EOS', '1' if (EVERYTHING and TREZOR_MODEL != "R") else '0'),
+    # TODO SOL: maybe "and TREZOR_MODEL != "R"" should be included?
```
Not from our side. Comment removed in https://github.com/trezor/trezor-firmware/pull/3359/commits/eeb91b764495f4e3082936c29f7eb0743166153b.

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+
+    is_multisig: bool = False
+    multisig_signers: list[Account] | None = None
+
```
[Changes](https://github.com/trezor/trezor-firmware/commit/728b7c796b76f8446a44860576bb543df9ca402c)
[Changes](https://github.com/trezor/trezor-firmware/commit/816defa398c36fb273a2e9d6f280d023318b0ad0)

albi84 commented on core/src/apps/solana/transaction/parse.py:
```diff
@@ -0,0 +1,227 @@
+from trezor.crypto import base58
+from trezor.wire import ProcessError
+
+from apps.common.readers import read_compact_size, read_uint32_le, read_uint64_le
```
[Changes](https://github.com/trezor/trezor-firmware/commit/7b115146d5656998e8a6f533a4b785f7caf5f18b)

albi84 commented on core/src/apps/solana/sign_tx.py:
```diff
@@ -0,0 +1,142 @@
+    from .ui import show_final_confirmation
+
+    address_n = msg.address_n
+    serialized_tx = msg.serialized_tx
```
[Changes](https://github.com/trezor/trezor-firmware/commit/5b5997f86b86940c3962232609c6c326764d9052)

albi84 commented on core/src/apps/solana/README.md:
```diff
@@ -0,0 +1,211 @@
+
+### Compact-u16 format
+
+A compact-u16 is a multi-byte encoding of 16 bits. The first byte contains the lower 7 bits of the value in its lower 7 bits. If the value is above 0x7f, the high bit is set and the next 7 bits of the value are placed into the lower 7 bits of a second byte. If the value is above 0x3fff, the high bit is set and the remaining 2 bits of the value are placed into the lower 2 bits of a third byte.
```
[Changes](https://github.com/trezor/trezor-firmware/commit/49446288563136d31b349d6d6178b7dd4c92a453)

albi84 commented on core/src/apps/solana/transaction/parse.py:
```diff
@@ -0,0 +1,227 @@
+        raise ProcessError(message)
+
+
+def parse_header(serialized_tx: BufferReader) -> tuple[bool, int, int, int, int]:
```
Do we really want to move this out from parse.py? Our intention was to keep all parsing function in one file.

matejcik replied:
thing is, `Transaction.__init__` is arguably a "parse function". so to me it would make much more sense to move this and friends ("parse addresses" "parse instructions" "parse lookup tables") and only keep things like "parse pubkey", "parse enum", that actually decode a certain data type

albi84 replied:
[Changes](https://github.com/trezor/trezor-firmware/commit/bf34dd0275f7838ce35b3448d987fbf916314553)

albi84 commented on core/src/apps/solana/transaction/instructions.py:
```diff
@@ -0,0 +1,5878 @@
+MEMO_PROGRAM_ID = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
+MEMO_LEGACY_PROGRAM_ID = "Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo"
+
+SYSTEM_PROGRAM_ID_INS_CREATE_ACCOUNT = 0
```
Actually yes, the Compute Budget related constants are used during the fee calculation.

albi84 commented on core/src/apps/solana/sign_tx.py:
```diff
@@ -0,0 +1,142 @@
+                instruction.instruction_id
+                == COMPUTE_BUDGET_PROGRAM_ID_INS_SET_COMPUTE_UNIT_PRICE
+            ):
+                unit_price = instruction.lamports
```
According to the Solana guys: [Compute budget instruction must be in the 1st 3 instructions (avoid nonce marker), otherwise ignored](https://github.com/solana-labs/solana/blob/647aa926673e3df4443d8b3d9e3f759e8ca2c44b/sdk/src/compute_budget.rs#L130-L132)
And multiple instances of that instruction _could_ exist but only the first one is respected, and generally it’s the first instructions.


[Changes](https://github.com/trezor/trezor-firmware/commit/c273d4358d4e80e2773fae78043e14c5b74f0b12)

matejcik replied:
if someone places two instructions, and the first sets unit price to 0 and the second to 10000, the fee calculation will be wrong here.
i'm afraid you'll need to set the `unit_limit` and `unit_price` to `None` first, to correctly detect which instruction set it first (contrary to that previous review comment from me)

albi84 commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,121 @@
+            ) = parse_address_lookup_tables(serialized_tx)
+
+        if serialized_tx.remaining_count() != 0:
+            raise ProcessError("Invalid transaction")
```
[Chnages](https://github.com/trezor/trezor-firmware/commit/c3df68d37ae11fb834a2aa078ed3441442c60662)

albi84 commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,121 @@
+from typing import TYPE_CHECKING
```
[Changes](https://github.com/trezor/trezor-firmware/commit/5422da83f8f9cabf036f2b2fe6db11c62e121c8d)

albi84 commented on core/src/apps/solana/transaction/instructions.py.mako:
```diff
@@ -0,0 +1,232 @@
+# generated from __init__.py.mako
```
[Changes](https://github.com/trezor/trezor-firmware/commit/09ec4f85dd533ee74db8bc11c5d0845afbf0255c)

albi84 commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    await confirm_properties(
+        "confirm_transaction",
+        "Confirm transaction",
+        [
```
[Changes part 1](https://github.com/trezor/trezor-firmware/commit/e7cc76ac870c2d61d93fcf82ae51db9d74c81964)

albi84 replied:
[Changes part 2](https://github.com/trezor/trezor-firmware/commit/066b6587fd32f4502ef5f17397e44967f4ea0c5c)

albi84 commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    signer_public_key: bytes,
+) -> None:
+    from trezor.enums import ButtonRequestType
+    from trezor.ui.layouts import confirm_metadata, confirm_properties
```
[Changes](https://github.com/trezor/trezor-firmware/commit/f3d4ffbb3a3ff7a8d05df02167bd2bdd7826980f)

albi84 commented on core/src/apps/solana/README.md:
```diff
@@ -0,0 +1,211 @@
+
+The process of generating the instruction builder involves the utilization of a [mako](https://www.makotemplates.org/) template engine, which is responsible for dynamically populating and transforming the data from `programs.json` into a Python script. This script, referred to as `instructions.py,` becomes an integral part of the software as it contains the Instruction builder function, facilitat ...
+
+The `programs.json` template file acts as a central reference point, allowing developers to easily update and adapt the software's behavior without modifying the core codebase. By editing the `programs.json`` file and regenerating the `instructions.py` script, developers can efficiently accommodate new Solana programs, modify existing ones, or make adjustments to account templates and parameter t ...
```
[Changes](https://github.com/trezor/trezor-firmware/commit/d2cbf7d927fab4685043fe8942d271d38b3dac2e)

albi84 commented on core/src/apps/solana/README.md:
```diff
@@ -0,0 +1,211 @@
+
+## Solana app basic concept and software architecture
+
+The software architecture designed for signing Solana messages utilizes a system of parser functions to construct objects from various classes. At its core is the Transaction class, serving as the higher-level entity responsible for encapsulating all transaction-related information. This includes a list of instructions derived from the Instruction class. This structure can seamlessly handle both  ...
```
[Changes](https://github.com/trezor/trezor-firmware/commit/b2051de5b107a12ca1b967906970a5c2d6f47d28)

albi84 commented on core/src/apps/solana/constants.py:
```diff
@@ -0,0 +1,11 @@
+ADDRESS_SIG = const(0)
+ADDRESS_SIG_READ_ONLY = const(1)
+ADDRESS_READ_ONLY = const(2)
+ADDRESS_RW = const(3)
```
[Changes](https://github.com/trezor/trezor-firmware/commit/b3fa8be4109d80ee044e43857233e29723a637d1)

albi84 commented on core/src/apps/solana/sign_tx.py:
```diff
@@ -0,0 +1,142 @@
+    node = keychain.derive(address_n)
+    signer_public_key = seed.remove_ed25519_prefix(node.public_key())
+
+    transaction: Transaction = Transaction(BufferReader(serialized_tx))
```
[Changes](https://github.com/trezor/trezor-firmware/commit/5cd654840fb93d797e168256425b2b019d15734e)

albi84 commented on core/src/apps/solana/transaction/parse.py:
```diff
@@ -0,0 +1,227 @@
+
+    addresses: list[Address] = []
+    for i in range(num_of_addresses):
+        assert_cond(ADDRESS_SIZE <= serialized_tx.remaining_count())
```
[Changes](https://github.com/trezor/trezor-firmware/commit/fac20cd0eb2d63f906bf140f61b44f97e0d5d620)

albi84 commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,145 @@
+        if attr in self.parsed_accounts:
+            return self.parsed_accounts[attr]
+
+        raise AttributeError(f"Attribute {attr} not found")
```
[Changes](https://github.com/trezor/trezor-firmware/commit/f7a3e40643b431ebbd2bcd2cc1f1e6308f114f25)

gabrielKerekes commented on core/src/apps/solana/types.py:
```diff
@@ -0,0 +1,62 @@
+    def __init__(self, name: str, type: str, optional: bool):
+        self.name = name
+        self.type = type
+        self.optional = optional
```
I tried that here -> https://github.com/trezor/trezor-firmware/pull/3359/commits/d81ddf5588c65d91acc3cc0537deb2e35464750f. Which parsing function gets used for which type is now actually defined in `programs.json` - I'm not 100% sure if it's a good idea to go this far, so please let me know what you think.

gabrielKerekes commented:
@matejcik all the review comments should be addressed. Could you please take a look? There's also the tests which are waiting for a review 🙏

cac-99 commented:
+1

matejkriz commented:
Checking the Solana send flow in Suite I found a big inconsistency. 

A) Recepient address confirmation and Amount sent confirmation are in reversed order. 
Technically it's possible to swap it in Suite, but all the other coins have it Address first and then amount, so it would be better to keep it the same for Solana if possible.

curren state:

---
1. step:
<img width="665" alt="image" src="https://github.com/trezor/trezor-firmware/assets/3729633/0a369a35-8f14-4ba0-bbee-7d0dca20f2ad">

---
2. step:
<img width="995" alt="image" src="https://github.com/trezor/trezor-firmware/assets/3729633/f7a38d60-acd2-41ee-83cc-20389543b04a">
---



B)  Last step (2 screens) is completely different from all other coins. Are Blockhash and Signer address something that user can and should verify? If so, it would be up to the Suite to guide the user. If not, Firmware screens should be changed, I guess.
<img width="1019" alt="image" src="https://github.com/trezor/trezor-firmware/assets/3729633/0f7e6f8e-28a9-438e-9b5c-0dc58589514e">
<img width="1012" alt="image" src="https://github.com/trezor/trezor-firmware/assets/3729633/11149c5e-57eb-49d9-8028-557b530586a3">
cc @Hannsek

matejkriz commented:
I wasn't able to confirm and finalize the send transaction, HOLD TO CONFIRM button was acting weird on my device, see the video. I've tried HOLD TO CONFIRM button in other flows and it works just fine on the same device. But I haven't tried another device, so it probably could be the touchscreen.
https://github.com/trezor/trezor-firmware/assets/3729633/0647f9b2-cd30-48cd-92b5-627f918aa7ea

matejcik commented on core/src/apps/solana/transaction/programs.json:
```diff
@@ -0,0 +1,3270 @@
+{
+  "programs": [
```
a very nice thing to have here would be a generator of a text file in very roughly the following format:

```
## System Program

Instruction: Create Account
1. Create Account:  (account) new_account
2. Deposit:         (lamports) lamports
3. From:            (account) funding_account
(not shown)         (u64) space
(not shown)         (authority) owner
```

so that (a) this description is auto-generated, CI checked, and (b) we have a straightforward ability to check which properties are shown and which aren't

i won't hold _this_ PR on it but please add it soon

matejcik commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,141 @@
+        self.accounts = accounts
+
+        if self.is_instruction_supported:
+            self.parsed_data = self.parse_instruction_data(
+                instruction_data, property_templates
+            )
+
+            self.parsed_accounts = self.parse_instruction_accounts(
+                accounts, accounts_template
+            )
```
these must not be conditional, otherwise you're not initializing `parsed_data` and `parsed_accounts`. the functions parse_instructions_x will cleanly handle the empty list of property/account templates, no?

gabrielKerekes replied:
In `parse_instruction_data` we verify that the whole buffer was used. This can't be done with unknown instructions so no, the functions don't handle empty templates cleanly. I think the best way to approach is simply add an `else` to the whole block which would initialize all three attributes - https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-8494caa0738cc8ad43c528cd82094529785133f6a2d7c4747c6ab7a91b6a034eR114.

matejcik commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,141 @@
+                accounts, accounts_template
+            )
+
+            self.multisig_signers = accounts[len(accounts_template) :]
+
+            if self.multisig_signers and not supports_multisig:
+                raise ValueError  # Multisig not supported
```
...and this should have an `else: self.multisig_signers = []`

gabrielKerekes replied:
As mentioned above - the three attributes are initialized in the `else` not just `multisig_signers`: https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-8494caa0738cc8ad43c528cd82094529785133f6a2d7c4747c6ab7a91b6a034eR114

matejcik commented on core/src/apps/solana/transaction/instruction.py:
```diff
@@ -0,0 +1,141 @@
+    @classmethod
+    def is_type_of(cls, ins: Any) -> TypeGuard[Self]:
+        # gets overridden in `instructions.py` `FakeClass`
+        return False
```
`raise NotImplementedError` would be a slightly better choice

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-8494caa0738cc8ad43c528cd82094529785133f6a2d7c4747c6ab7a91b6a034eR144

matejcik commented on core/src/apps/solana/ui/format.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def format_lamports(_: Instruction, value: int) -> str:
+    from trezor.strings import format_amount
```
small nit: `from trezor.strings import format_amount, format_timestamp` should be top-level import.

(the rules for function-local imports are somewhat unclear, but generally a single top-level import is preferred in case it replaces two or more function-local imports)

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-0ed8070728ed70f5cd08bb7f7f5e4a53db0e6fff3002f21ca465408bfb9dc824R3

matejcik commented on core/src/apps/solana/ui/format.py:
```diff
@@ -0,0 +1,42 @@
+def format_token_amount(instruction: Instruction, value: int) -> str:
+    from trezor.strings import format_amount
+
+    decimals = instruction.decimals if instruction.decimals is not None else 0
```
does this always work, i.e., is the `instruction` guaranteed to have `decimals` ?

gabrielKerekes replied:
No, it's not really guaranteed although it is true now - any instruction using `format_token_amount` does have a decimals attr although it's not represented in the code.

I'm not sure what the best way to fix this is. One option is to return `None` from from `Instruction.__getattr` if the attr is not found. Then the current condition wouldn't cause problems anymore although that's IMO a bug prone solution.

Another is to simply wrap this line in a `try/except` and set `decimals = 0` in the `except` block. This could also be extracted into `Instruction.get_attr_with_default` for example (but extracting it would also require the attribute to be referenced by a string).

Another is to do `if "decimals" in instruction and instruction.decimals is not None` instead of the current condition - although I don't like the attribute being referenced by string.

I don't know of a simple (or perhaps any) way to represent this also in the type system somehow.

matejcik replied:
like `assert hasattr(instruction, "decimals")  # enforced by build_solana_templates.py`

gabrielKerekes replied:
> in which case would instruction.decimals exist but be None?

The `is None` check was there because I thought that it would work as a check for the instruction not having the decimals - the code is wrong (also in my comment), we don't expect decimals to exist and be None.

> and, do we ever expect to have an instruction that uses format_token_amount but does not have decimals ?

No.

> i'm leaning towards keeping this code as-is, but adding a static check to the template generator, to the tune of "if using format_token_amount, then check that there's also a decimals field"? (and referencing it in a comment here)

Could work. I'll take a look.

gabrielKerekes replied:
Templates updated in https://github.com/trezor/trezor-firmware/pull/3359/commits/8a17e693a1ef0d1fc865458857c600adc3b4a2a3 and code updated in https://github.com/trezor/trezor-firmware/pull/3359/commits/4d9f08502322c4dcddfa096e7651fef99241ffc7.

matejcik commented on core/SConscript.firmware:
```diff
@@ -72,6 +72,7 @@ CPPDEFINES_MOD += [
     ('USE_CARDANO', '1' if EVERYTHING else '0'),
     ('USE_NEM', '1' if (EVERYTHING and TREZOR_MODEL != "R") else '0'),
     ('USE_EOS', '1' if (EVERYTHING and TREZOR_MODEL != "R") else '0'),
+    ('USE_SOLANA', '1' if EVERYTHING else '0'),
```
this is only relevant for conditional compilation of trezor-crypto, which this PR doesn't modify. so please remove from here and sconscript.unix

gabrielKerekes replied:
Removed https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-2bd28a262d65469606d7b9a80760c5d905e26ab8fe8406704173eaac218f56b6L75.

matejcik commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,246 @@
+    return f"#{unharden(account_index) + 1}"
+
+
+async def show_confirm(
```
rename to `confirm_instruction` (naming scheme is that `show_xyz` only show information and proceed, `confirm_xyz` raise an exception if the user aborts (which is the case here))

dtto for `show_*` below

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-83f3741450ac075bc7410d882727013da398123a64c271336294fb764e172168R27

matejcik commented on core/src/apps/solana/ui/__init__.py:
```diff
@@ -0,0 +1,246 @@
+        )
+
+
+def get_address_type(address_type: int) -> str:
```
```suggestion
def get_address_type(address_type: AddressType) -> str:
```

matejcik commented on tests/device_tests/solana/test_address.py:
```diff
@@ -0,0 +1,42 @@
+    "solana/get_address.json",
+)
+def test_solana_get_address(client: Client, parameters, result):
+    client.init_device(new_session=True)
```
why do you call init_device here?

gabrielKerekes replied:
I think I copied the tests from Cardano which call `init_device` to set `derive_cardano=True` so I just removed the parameter here instead of removing the whole function. Removed from all three test functions in https://github.com/trezor/trezor-firmware/pull/3359/commits/5cc7eab0731883cd9af6fb2e3eba1398a667f12e#diff-bcf8436b9cb87e3509d72fce300324b5587a11b27b4787d05b40ea3a8ec57ce8L36.

matejcik commented on tests/device_tests/solana/construct/instructions.py:
```diff
@@ -0,0 +1,1443 @@
+
+
+AssociatedTokenAccountProgram_Instruction = Switch(
+    lambda this: this.instruction_id,
```
```suggestion
    this.instruction_id,
```

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/5cc7eab0731883cd9af6fb2e3eba1398a667f12e#diff-3a06ea4817ce485da542a9c3f705d02ecd5f1e7f8da16375d9766218503d4e07R90

matejcik commented on tests/device_tests/solana/construct/instructions.py:
```diff
@@ -0,0 +1,1443 @@
+    ),
+    "data"
+    / CompactStruct(
+        "instruction_id" / InstructionIdAdapter(GreedyBytes),
```
you can use `InstructionId` in all those cases, which you helpfully already have defined :)

gabrielKerekes replied:
https://github.com/trezor/trezor-firmware/pull/3359/commits/5cc7eab0731883cd9af6fb2e3eba1398a667f12e#diff-3a06ea4817ce485da542a9c3f705d02ecd5f1e7f8da16375d9766218503d4e07R80

matejcik reviewed (changes requested):
sorry for taking so long on the review.
i have multiple comments still but they are now mostly minor local changes.

good job overall :100: 

the construct code for tests did shape up nicely in the end.

one thing I would like to see is the documentation generator from programs.json -- but it does not need to be part of this PR

matejcik commented:
@mmilata please take a look at the code in `solana/ui/__init__py` with regards to @matejkriz's comments, and advise what to do here.

I believe we can make some things better, like moving some of the final confirm information behind an (i) button

mmilata commented:
The HOLD TO CONFIRM issue is weird, I haven't been able to reproduce it using device tests but it's unlikely to be hardware issue if it happens only in the solana flow. @matejkriz can you please post exact instructions to reproduce?

Regarding the final confirmation screen, please take a look at 3f8db0878d8420bd15304bb9fa2e2d5ac8f4c6a5 for an example of how it can look similar to other coins. Though I'm not sure which info belongs on the main screen and which on the info screen.

![emu00000000](https://github.com/trezor/trezor-firmware/assets/85857/aa1e48a6-32a0-4092-aa45-504aeda2f729) ![emu00000001](https://github.com/trezor/trezor-firmware/assets/85857/9765bf2a-3fca-44f5-863d-c916d9053505) ![emu00000002](https://github.com/trezor/trezor-firmware/assets/85857/88a7861d-dcb0-45ef-844b-6e4aacac32f0)

matejcik commented:
one more nit, @gabrielKerekes @albi84 please move `ui/__init__.py` to `layout.py` (this is the name other apps are using) and `ui/format.py` to just `format.py` (no need to nest it anywhere)

gabrielKerekes commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,214 @@
+        Combine accounts from transaction's accounts field with accounts from address lookup tables.
+        Instructions reference accounts by index in this combined list.
+        """
+        accounts: list[Account] = []
+        for address in self.addresses:
+            accounts.append(address)
+
+        for rw_address in self.address_lookup_tables_rw_addresses:
+            accounts.append(rw_address)
+        for ro_address in self.address_lookup_tables_ro_addresses:
+            accounts.append(ro_address)
+
+        return accounts
+
```
Unfortunately no. Pyright then complains about `Operator "+" not supported for types "list[Address]" and "list[AddressReference]"` - I don't know of a way to circumvent that other than the function we have now.

matejcik replied:
I see the problem. We would need to explicitly `cast` to `list[Account]` (a strictly correct thing to do), but we don't have `cast()` here. So please just `# type: ignore` it.

gabrielKerekes replied:
Ignored in https://github.com/trezor/trezor-firmware/pull/3359/commits/78d5c99cd1f1c1b12e0e4bafebb72308733738d3. I tried that before but I wasn't able to figure the correct way to ignore it 🙈

gabrielKerekes commented on core/src/apps/solana/transaction/parse.py:
```diff
@@ -0,0 +1,54 @@
+        shift += 7
+        if B & 0b10000000 == 0:
+            return value
+    return value
```
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-852762cb5345206c8b686136d9adf024f2385c3f2f46444c25f7748ec642ee4fR17

gabrielKerekes commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,214 @@
+            # id in the data for instruction id 0 but they include it for the other instructions.
+            # Instructions with such optional instruction id also don't contain any other instruction
+            # data (otherwise parsing would be impossible).
+            if data_length < instruction_id_length:
```
Actually no, because e.g. the Memo program only has a single instruction so it doesn't include instruction id in the data but it does contain other instruction data.

The comment was misleading so I updated it in https://github.com/trezor/trezor-firmware/pull/3359/commits/07a33b0ea99674b761bcda99186dac23047148cd and then once more in https://github.com/trezor/trezor-firmware/pull/3359/commits/5ed1636c7efc3d45f30d3752dff67fadf9a38f30.

gabrielKerekes commented on core/src/apps/solana/transaction/__init__.py:
```diff
@@ -0,0 +1,214 @@
+    def __init__(self, serialized_tx: bytes) -> None:
+        self.instructions = []
+        self.address_lookup_tables_rw_addresses = []
+        self.address_lookup_tables_ro_addresses = []
```
https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-774cb9f6449b3d72ef6bb889b99659d934897992c7e52107409afebfed845406L33.

gabrielKerekes commented:
> please move ui/__init__.py to layout.py (this is the name other apps are using) and ui/format.py to just format.py (no need to nest it anywhere)

Done in https://github.com/trezor/trezor-firmware/pull/3359/commits/caf89cc655e2588e9c1f874ed5aab830cccb6835.

I also noticed that we didn't use `format_token_amount` for some instructions so I added that as well in https://github.com/trezor/trezor-firmware/pull/3359/commits/09cb9825256481fe659b784a755efdac3c622023.

gabrielKerekes commented on core/src/apps/solana/types.py:
```diff
@@ -0,0 +1,75 @@
+
+    from .transaction import Instruction
+
+    Address = tuple[bytes, int]
```
Thanks to switching to `AddressType` I noticed a bug with showing unknown instructions - we would incorrectly show accounts from lookup tables. I will fix this bug along with the type change in about an hour or so.

gabrielKerekes replied:
Fixed in https://github.com/trezor/trezor-firmware/pull/3359/commits/e996e0de4d0b91f2fa6f17d978f915491b6122a4.

I also noticed that unknown instruction data wasn't being displayed properly because the underlying type was `memoryview` instead of `bytes` so I changed the instruction data's type to be explicitly `memoryview` in https://github.com/trezor/trezor-firmware/pull/3359/commits/c06b13c6821ed8f01ffb94d88783cf2b4e802fed. Apparently `memoryview` can be assigned to `bytes` without pyright complaining but the `confirm_properties` call works with `bytes` only.

gabrielKerekes commented:
> Regarding the final confirmation screen, please take a look at https://github.com/trezor/trezor-firmware/commit/3f8db0878d8420bd15304bb9fa2e2d5ac8f4c6a5 for an example of how it can look similar to other coins. Though I'm not sure which info belongs on the main screen and which on the info screen.

Nice, thanks, I cherry-picked your commit. You guessed the information to be shown correctly 👍

matejcik reviewed (commented):
ACK from my side

it is time to generate the UI fixtures

matejcik commented:
having the fixtures now lets us see how the tweaks affect them.

gabrielKerekes commented:
Oh, actually we should probably first rewrite all the tests into the construct syntax, right?

matejcik commented:
i thought the tests are ok like this? json decoded into construct

gabrielKerekes commented:
> i thought the tests are ok like this? json decoded into construct

But only a sample of tests is written like that. All the tests in `sign_tx.tmp.json` should be rewritten.

mmilata commented:
Also feel free to cherry-pick 84c69461ca0b16d9be8d72ba0920af4fcfba3fb5 to keep `Blockhash:` title and the actual value on the same page (untested).

![emu00000001](https://github.com/trezor/trezor-firmware/assets/85857/62d6f83d-0612-45c5-8a8b-133b54de922c) ![emu00000002](https://github.com/trezor/trezor-firmware/assets/85857/eba24f06-332b-42fa-b213-b78c469af7d1)

gabrielKerekes commented:
As discussed offline - send SOL and send Token (+ create token account) flows updated in https://github.com/trezor/trezor-firmware/pull/3359/commits/4e857afa82d884963e13314543fe3c38902574b3. I updated the token flow as well as it was rather simple to do the way I handled it. Let me know what you think about it 🙏 

<details><summary>screenshots here</summary>

### Send SOL

![image](https://github.com/trezor/trezor-firmware/assets/7520735/31d28f5d-4c51-4583-8c96-445d9c6ca6bd)
![image](https://github.com/trezor/trezor-firmware/assets/7520735/18dafcf6-e751-47a8-9c82-746a0b3686ea)
![image](https://github.com/trezor/trezor-firmware/assets/7520735/9c4307a1-a1a0-4b53-b4ca-63b50e60f80f)

### Send token with account creation
![image](https://github.com/trezor/trezor-firmware/assets/7520735/53d34f68-c32c-46cb-883d-65f3c36f9d98)
![image](https://github.com/trezor/trezor-firmware/assets/7520735/120d56bb-148a-4890-9baf-7f664ea67494)
![image](https://github.com/trezor/trezor-firmware/assets/7520735/4fc7e711-7d69-454d-8e45-4b277fa42038)

The `(account will be created)` under the recipient address in token flow is still under discussion. The text will be hidden if the account is being created.
</details>

I'm still working on adding all the tests - I hoped to use the construct parser to build the JSON objects from serialised txs but apparently deserialising via Construct doesn't work properly so I was investigating and trying to fix that as well but I was not able to fix it in a reasonable time - so I'd probably leave it for now, add tests manually and we can discuss later if we should fix it - but I might need help doing that.

Hannsek commented:
Looking good, thanks! 👍🏻 Some small nitpicks:

- `Account` should be `Solana #1`
- `Blockhash` can be longer, no? So I'd rather put it on the separate screen.

gabrielKerekes commented:
> Blockhash can be longer, no? So I'd rather put it on the separate screen.

No, it's always 32 bytes and although the base58 encoding length varies it should always fit on that screen.

> Account should be Solana #1

Missed that, fixed in https://github.com/trezor/trezor-firmware/pull/3359/commits/9c5f105d32d209fa95d7996fd193c8db1a7b65d4.

> The text will be hidden if the account is already created, right?

Right, I updated my comment, sorry.

Hannsek commented:
> > Blockhash can be longer, no? So I'd rather put it on the separate screen.

> No, it's always 32 bytes and although the base58 encoding length varies it should always fit on that screen.

Ok, then we can keep it for now. Once we will implement [this](https://github.com/trezor/trezor-firmware/issues/2353) it will have to be moved onto the other screen.

mmilata commented:
Please one more cherry-pick:pray:  da4408ae1

Changes the scrollbar orientation for transaction summary details, as requested on slack:

![emu00000000](https://github.com/trezor/trezor-firmware/assets/85857/8b8fc9b9-e291-404a-9143-28c60f453f7a) ![emu00000001](https://github.com/trezor/trezor-firmware/assets/85857/a3b40bd1-de4a-4168-96c8-746539150a1b)

mmilata commented:
@Hannsek  We should probably still merge it because it affects all coins and according to figma the scrolling should be horizontal everywhere, right? Probably didn't catch this because most coins only have one screen worth of info. I'll open separate PR.

matejcik commented:
is there anything pending in this PR? we have some outstanding modifications in the pipeline but they are not necessary now.
@Hannsek @gabrielKerekes 

if not, @gabrielKerekes please fix the conflicts, squash fixups, generate UI fixtures, so that we can get this in

gabrielKerekes commented:
I'm still working on adding all the tests. There's possibly one bug I need to fix regarding optional params. I also want to check your updated Construct stuff, but that can be done after this PR is merged. I'll post an update around noon.

Hannsek commented:
Nothing from my side. :)

gabrielKerekes commented:
I'm still finalizing. All tests passed. I'm replacing `is_included_if_zero` - I thought it's done, but the tests are now failing. Will report soon.

gabrielKerekes commented:
In https://github.com/trezor/trezor-firmware/pull/3359/commits/12230171c4826d8214c513ce59f98ad9ec0b63ff I fixed a bug with optional types - e.g. in the StakeProgram.SetLockup instruction - the parameters are prefixed with a byte indicating whether the value is included or not 🙈 

In https://github.com/trezor/trezor-firmware/pull/3359/commits/9c4abd72a1d35f284416ddba6dc7ffc5dc26df12 I added tests for all the instructions.
 
 In https://github.com/trezor/trezor-firmware/pull/3359/commits/548ff86c1137abe53dcefe820aa98758115672a6 I got rid of `is_included_if_zero`. I also partially applied Construct updates from https://github.com/trezor/trezor-firmware/commit/619e8eae1bae3390782b87b8a30dd7e5f7d57203#diff-c37f39a3e4fbe3bcc17153790015d971bb685352179ddc79a0682336d17bd7d1. [This](https://github.com/trezor/trezor-firmware/pull/3359/commits/548ff86c1137abe53dcefe820aa98758115672a6#diff-774cb9f6449b3d72ef6bb889b99659d934897992c7e52107409afebfed845406R125) had to be brought (not sure where we removed it anymore) because of AssociatedTokenAccountProgram.Create.
 
 I didn't squash and rebase yet, so you could review the new updates. I'm heading off now, so I'll fix new comments if any in the evening/tomorrow morning and squash and rebase then as well.
 
I think the Construct code should be updated further to match https://github.com/trezor/trezor-firmware/commit/619e8eae1bae3390782b87b8a30dd7e5f7d57203#diff-c37f39a3e4fbe3bcc17153790015d971bb685352179ddc79a0682336d17bd7d1 but we can do that after this PR is merged. The FW mako template could then probably be updated as well.

gabrielKerekes commented:
I made one more update which basically finalises the Construct/template updates according to your suggestions https://github.com/trezor/trezor-firmware/pull/3359/commits/02e6766e2daabb0099c954ecba4a0cd89673aa20. So the outstanding things to review/resolve are in https://github.com/trezor/trezor-firmware/pull/3359#issuecomment-1833900707 + https://github.com/trezor/trezor-firmware/pull/3359/commits/02e6766e2daabb0099c954ecba4a0cd89673aa20 and the base58 lib issues should probably be discussed as well (https://github.com/trezor/trezor-firmware/pull/3359#discussion_r1411732442). cc @matejcik

gabrielKerekes commented:
> please squash, rebase on current main (so that we get github actions for the PR), and also generate fixtures for TS3
(same steps as for TT, except emulator must be built with TREZOR_MODEL=R)

Done, although I had to make a rather temporary fix to Trezor R in https://github.com/trezor/trezor-firmware/pull/3359/commits/bc2e4133f46dd8c5b39143001415abee0eea16ed.

gabrielKerekes commented:
In https://github.com/trezor/trezor-firmware/pull/3359/commits/5922988cade990d24550c54469eb08ed69618824 I modified `confirm_value` so it doesn't send unnecessary messages to Suite - sending multiple messages would mess up the steps in transaction review dialog. However, now when the user goes to info and then back, the recipient address is not displayed - there's some redraw issue although `should_show_more_layout.request_complete_repaint()` is called. Do you have any idea what might be wrong? cc @grdddj 🙏 

To test: (the issue is only on Trezor R)
```
poetry run trezorctl solana sign-tx -n "m/44'/501'/0'/0'" -a additional_info.json  0100040700d1699dcb1811b50bb0055f13044463128242e37a463b52f6c97a1f6eef88ad9bdc8c11af6a471f7373e52c917aac6304c71796b97b47350d46cf4c54bae9d9fdd530a73d5e74613b17aaeb9346f9fc3c5fd6a7ad126b269e3e73454e2a0b1b000000000000000000000000000000000000000000000000000000000000000081432588d91c8e02d613baa94e05994e78467cf5a8821d8f4c2e373e8b9b56ef8c97258f4e2489f1bb3d1029148e0d830b5a1399daff1084048e7bd8dbe9f85906ddf6e1d765a193d9cbe146ceeb79ac1cb485ed5f5b37913a8cf5857eff00a91aea57c9906a7cad656ff61b3893abda63f4b6b210c939855e7ab6e54049213d020506000201040306000604010402000a0c0b0000000000000009
```
Additional info file:
```
{
  "token_accounts_infos": [
    {
      "base_address": "BVRFH6vt5bNXub6WnnFRgaHFTcbkjBrf7x1troU1izGg",
      "token_program": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
      "token_mint": "9hayiPmEobVfiTbw5R91StWeQzw9EJGfswLH5o33UDAW",
      "token_account": "J5rhFGUkeoHVnCvMyqWq1XPjfU1G1hsTh9tTQtST2out"
    }
  ]
}
```

gabrielKerekes commented:
> However, now when the user goes to info and then back, the recipient address is not displayed

Fixed - by moving the layout creation into while loop. https://github.com/trezor/trezor-firmware/pull/3359/commits/2dfe21a5cc447291728dad8d70f76e8959baa2bc

So now I think we should be ready for a final review or merging. cc @matejcik I'm regenerating the ui fixtures again, but I don't expect them to change.

gabrielKerekes commented on core/src/trezor/ui/layouts/tr/__init__.py:
```diff
@@ -966,29 +966,50 @@ async def confirm_value(
+                            )
+                        )
+                    )
+                except ActionCancelled:
```
👍 https://github.com/trezor/trezor-firmware/pull/3359/commits/3465e21d09d887280c6d8ea147d228dac307bbe0

gabrielKerekes commented:
Changelog added in https://github.com/trezor/trezor-firmware/pull/3359/commits/583b07ebcfebc5baff0ff678719edae76f020a34.

See diff here for changes to `build_solana_templates.sh` - the problem was (AFAICT) that the `tmp` folder lives outside of the core directory and so some of the imports were meant to have a different order because of that. I fixed by creating and deleting a file in `./core/tools/` 🤷 I couldn't think of anything better. Adding `fmt: skip` or `fmt: off` and `fmt: on` comments to the template file didn't work.

gabrielKerekes commented:
I rebased on top of current `main` to resolve conflicts.

matejcik reviewed (approved):
waiting for one last CI run

cac-99 commented:
🔥 🔥 🔥

bosomt commented:
**QA OK**

**Info:**
 - Suite version: desktop 23.12.2 (a4ee73cb452255eae83296cf1685e790d87c9eb0)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/23.12.2 Chrome/118.0.5993.129 Electron/27.0.4 Safari/537.36
 - OS: MacIntel
